alvr/sockets/src/stream_socket/udp.rs

```diff
@@ -0,0 +1,270 @@
+    mem::{self, MaybeUninit},
+    net::{IpAddr, UdpSocket},
+    ptr,
+    time::Duration,
+};
+
+pub const SHARD_PREFIX_SIZE: usize = mem::size_of::<u16>() // stream ID
+    + mem::size_of::<u32>() // packet index
+    + mem::size_of::<u32>() // shards count
+    + mem::size_of::<u32>(); // shards index
+
+fn socket_peek(socket: &mut Socket, buffer: &mut [u8]) -> ConResult<usize> {
+    #[cfg(windows)]
+    const FLAGS: c_int = 0x02 | 0x8000; // MSG_PEEK | MSG_PARTIAL
+    #[cfg(not(windows))]
+    const FLAGS: c_int = 0x02 | 0x20; // MSG_PEEK | MSG_TRUNC
+
+    let buffer =
+        MaybeUninitSlice::new(unsafe { &mut *(ptr::from_mut(buffer) as *mut [MaybeUninit<u8>]) });
+    // NB: Using the non vectored call doesn't seem to work
+    Ok(socket
+        .recv_vectored_with_flags(&mut [buffer], FLAGS)
+        .handle_try_again()?
+        .0)
+}
+
+// Create tokio socket, convert to socket2, apply settings, convert back to tokio. This is done to
+// let tokio set all the internal parameters it needs from the start.
+pub fn bind(
+    port: u16,
+    dscp: Option<DscpTos>,
+    send_buffer_bytes: SocketBufferSize,
+    recv_buffer_bytes: SocketBufferSize,
+) -> Result<UdpSocket> {
+    let socket = UdpSocket::bind((LOCAL_IP, port))?.into();
+
+    crate::set_socket_buffers(&socket, send_buffer_bytes, recv_buffer_bytes).ok();
+
+    crate::set_dscp(&socket, dscp);
+
+    Ok(socket.into())
+}
+
+pub fn connect(socket: &UdpSocket, peer_ip: IpAddr, port: u16, timeout: Duration) -> Result<()> {
+    socket.connect((peer_ip, port))?;
+    socket.set_read_timeout(Some(timeout))?;
+
+    Ok(())
+}
+
+pub struct MultiplexedUdpWriter {
+    inner: UdpSocket,
+    max_packet_size: usize,
+}
+
+impl MultiplexedSocketWriter for MultiplexedUdpWriter {
+    fn payload_offset(&self) -> usize {
+        SHARD_PREFIX_SIZE
+    }
+
+    fn send(&mut self, stream_id: u16, packet_index: u32, buffer: &mut Vec<u8>) -> Result<()> {
+        let max_shard_size = self.max_packet_size - SHARD_PREFIX_SIZE;
+        let payload_size = buffer.len() - SHARD_PREFIX_SIZE;
+        // rounding up:
+        let shards_count = payload_size.div_ceil(max_shard_size);
+
+        for shard_idx in 0..shards_count {
+            // this overlaps with the previous shard, this is intended behavior and allows to
+            // reduce allocations
+            let shard_start_position = shard_idx * max_shard_size;
+            let shard_size = usize::min(max_shard_size, payload_size - shard_start_position);
+
+            let shard_view = &mut buffer[shard_start_position..][..SHARD_PREFIX_SIZE + shard_size];
+
+            shard_view[0..2].copy_from_slice(&stream_id.to_le_bytes());
+            shard_view[2..6].copy_from_slice(&packet_index.to_le_bytes());
+            shard_view[6..10].copy_from_slice(&(shards_count as u32).to_le_bytes());
+            shard_view[10..14].copy_from_slice(&(shard_idx as u32).to_le_bytes());
+
+            self.inner.send(shard_view)?;
+        }
```

Comment:
Honestly this would be a **great** place for an async send operation that just does all in one syscall

Reply:
What do you mean with this? It's to our advantage not to send the whole packet at once and instead interleave the shards from other threads. Practically this lowers the overall latency.